### PR TITLE
Adding gRPC Option to Client

### PR DESCRIPTION
**Pull request by olapiv**
Refactored quite a bit. The REST API worked, gRPC had weird runtime errors.

### CODE REVIEW

File: rondb/src/main/java/site/ycsb/db/RonDBClient.java [hunk 1]
```diff
@@ -25,103 +25,102 @@
...
-      final String fieldnameprefix = getProperties().getProperty(CoreWorkload.FIELD_NAME_PREFIX,
+
+      useRESTAPI = Boolean.parseBoolean(properties.getProperty(RONDB_USE_REST_API, "false"));
+      useGRPC = Boolean.parseBoolean(properties.getProperty(RONDB_USE_GRPC, "false"));
+      if (useRESTAPI && useGRPC) {
+        logger.error("cannot use both REST API and GRPC");
+        System.exit(1);
+      }
+
+      // It can apparently happen that the methods omit the parameter "fields", so
```
Comment (smkniazi) [inline rondb/src/main/java/site/ycsb/db/RonDBClient.java:86]: remove comment

Comment (smkniazi) [reply rondb/src/main/java/site/ycsb/db/RonDBClient.java:86]: wrong comment.

File: rondb/src/main/java/site/ycsb/db/RonDBClient.java [hunk 2]
```diff
@@ -25,103 +25,102 @@
...
    */
   public void cleanup() throws DBException {
-//    System.out.println("----------------> cleanup");
-//    RonDBRestClient.getClient().notifyAllBarriers();
-    synchronized (lock) {
-      if (connection != null) {
-        RonDBConnection.closeSession(connection);
-      }
-    }
+    ClusterJClient.cleanup();
```
Comment (smkniazi) [inline rondb/src/main/java/site/ycsb/db/RonDBClient.java:123]: why you have removed sync block

File: rondb/src/main/java/site/ycsb/db/clusterj/ClusterJClient.java [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+          e.printStackTrace();
+          System.exit(1);
+        }
+      }
+      connection.returnSession(session);
+    }
+  }
+
+  public static synchronized void cleanup() throws DBException {
+    if (connection != null) {
```
Comment (smkniazi) [inline rondb/src/main/java/site/ycsb/db/clusterj/ClusterJClient.java:83]: sync is needed here

File: rondb/src/main/java/site/ycsb/db/clusterj/table/UserTableHelper.java [hunk 4]
```diff
@@ -42,9 +42,8 @@ private UserTableHelper() {
 
   }
 
-  public static DynamicObject createDTO(ClassGenerator classGenerator, Session session,
-                                        String tableName, String keyVal,
-                                        Map<String, ByteIterator> values) throws Exception {
+  public static DynamicObject createDTO(ClassGenerator classGenerator, Session session, String tableName, String keyVal,
```
Comment (smkniazi) [inline rondb/src/main/java/site/ycsb/db/clusterj/table/UserTableHelper.java:45]: please avoid such changes. This is very annoying when you auto-format code using a different scheme. stick to the formatting scheme used by the file

File: rondb/pom.xml [hunk 5]
```diff
@@ -35,6 +35,30 @@ LICENSE file.
...
   </properties>
 
+  <build>
+    <plugins>
+      <plugin>
+        <groupId>org.xolstice.maven.plugins</groupId>
+        <artifactId>protobuf-maven-plugin</artifactId>
+        <version>0.6.1</version>
+        <configuration>
+          <!-- Check this for codegen on Mac OSX M1: https://github.com/grpc/grpc-java/issues/7690#issuecomment-772424454 -->
```
Comment (smkniazi) [inline rondb/pom.xml:45]: this breaks build on linux

File: rondb/src/main/java/site/ycsb/db/clusterj/table/UserTableHelper.java [hunk 6]
```diff
@@ -59,8 +58,7 @@ public static DynamicObject createDTO(ClassGenerator classGenerator, Session ses
     return persistable;
   }
 
-  private static void setFieldValue(DynamicObject persistable, String colName, byte[] value,
-                                    int lenght) {
+  private static void setFieldValue(DynamicObject persistable, String colName, byte[] value, int lenght) {
```
Comment (smkniazi) [inline rondb/src/main/java/site/ycsb/db/clusterj/table/UserTableHelper.java:61]: why would you do such a change?

**Review by smkniazi — COMMENTED** [review]
You need to fix two critical

1. compilation issue.
2. Sync 

Please stick to existing formatting scheme of src files

File: rondb/src/main/java/site/ycsb/db/clusterj/ClusterJClient.java (hunk 3 above)
Comment (olapiv) [reply rondb/src/main/java/site/ycsb/db/clusterj/ClusterJClient.java:83]: I use the keyword `synchronized`..?

File: rondb/src/main/java/site/ycsb/db/RonDBClient.java (hunk 2 above)
Comment (olapiv) [reply rondb/src/main/java/site/ycsb/db/RonDBClient.java:123]: It belongs to the ClusterJClient.

File: rondb/src/main/java/site/ycsb/db/RonDBClient.java (hunk 1 above)
Comment (olapiv) [reply rondb/src/main/java/site/ycsb/db/RonDBClient.java:86]: Remove or not? If it's wrong, why so?

File: rondb/pom.xml (hunk 5 above)
Comment (olapiv) [reply rondb/pom.xml:45]: Any suggestions? I run on Mac and am also not an expert in Maven.

File: rondb/src/main/java/site/ycsb/db/clusterj/table/UserTableHelper.java (hunk 4 above)
Comment (olapiv) [reply rondb/src/main/java/site/ycsb/db/clusterj/table/UserTableHelper.java:45]: I ran an auto formatter before I saw the `checkstyle.xml` file. I tried to revert what I can, looks like this slipped.

File: rondb/src/main/java/site/ycsb/db/clusterj/table/UserTableHelper.java (hunk 6 above)
Comment (olapiv) [reply rondb/src/main/java/site/ycsb/db/clusterj/table/UserTableHelper.java:61]: Autoformatter.

File: rondb/src/main/java/site/ycsb/db/clusterj/ClusterJClient.java (hunk 3 above)
Comment (smkniazi) [reply rondb/src/main/java/site/ycsb/db/clusterj/ClusterJClient.java:83]: `synchronized (lock)` and making the fn synchronized is not the same thing.

File: rondb/src/main/java/site/ycsb/db/RonDBClient.java (hunk 1 above)
Comment (smkniazi) [reply rondb/src/main/java/site/ycsb/db/RonDBClient.java:86]: this is an unnecessary comment and the contents are wrong

File: rondb/src/main/java/site/ycsb/db/clusterj/ClusterJClient.java (hunk 3 above)
Comment (smkniazi) [reply rondb/src/main/java/site/ycsb/db/clusterj/ClusterJClient.java:83]: use `synchronized (lock)` as used in the function above

File: rondb/src/main/java/site/ycsb/db/clusterj/table/UserTableHelper.java (hunk 4 above)
Comment (smkniazi) [reply rondb/src/main/java/site/ycsb/db/clusterj/table/UserTableHelper.java:45]: there are a lot of such changes. Please revert such changes

File: rondb/src/main/java/site/ycsb/db/clusterj/table/UserTableHelper.java (hunk 6 above)
Comment (smkniazi) [reply rondb/src/main/java/site/ycsb/db/clusterj/table/UserTableHelper.java:61]: Please revert such changes. The existing code is correctly formatted in this case

File: rondb/pom.xml (hunk 5 above)
Comment (smkniazi) [reply rondb/pom.xml:45]: I will try to fix it tomorrow

File: rondb/src/main/java/site/ycsb/db/clusterj/table/UserTableHelper.java (hunk 6 above)
Comment (olapiv) [reply rondb/src/main/java/site/ycsb/db/clusterj/table/UserTableHelper.java:61]: I just ran Checkstyle over it and apparently I didn't break anything. I'll continue running an auto formatter over the code that I write and in this case I removed an unnecessary newline. I personally found it easier to read this way. But this is a PR towards your repo anyways, so feel free to change anything back before it's merged to logicalclocks, where I'll rebase from again afterwards.

File: rondb/src/main/java/site/ycsb/db/clusterj/ClusterJClient.java (hunk 3 above)
Comment (olapiv) [reply rondb/src/main/java/site/ycsb/db/clusterj/ClusterJClient.java:83]: I don't think I agree. If the method is also static, then I believe it is the same thing.

https://www.geeksforgeeks.org/static-synchronization-in-java/

>  We are attaining a class-level lock such that only one thread will operate on the method.

File: rondb/src/main/java/site/ycsb/db/RonDBClient.java (hunk 1 above)
Comment (olapiv) [reply rondb/src/main/java/site/ycsb/db/RonDBClient.java:86]: Okay, but why did you write the following then?

```java
public Status read(String table, String key, Set<String> fields, Map<String, ByteIterator> result) {
    Set<String> fieldsToRead = fields != null ? fields : fieldNames;  // <-- surely this is unnecessary then?
    ...
```

**Pull request merged**